vllm/core/block/cpu_gpu_block_allocator.py

```diff
@@ -163,19 +170,22 @@ def fork(self, last_block: Block) -> List[Block]:
             List[Block]: A new list of blocks that shares the same memory as the
                 original sequence.
         """
-        allocator = self._block_ids_to_allocator[last_block.block_id]
+        block_id = last_block.block_id
+        assert block_id is not None
+        allocator = self._block_ids_to_allocator[block_id]
         return allocator.fork(last_block)
 
-    def get_num_free_blocks(self, device: Device) -> int:
+    def get_num_free_blocks(self, device: Optional[Device] = None) -> int:
```

Comment:
Imo this is a band-aid solution, it may be better to fix the inheritance so that it actually satisfies LSP.

Reply:
but yeah I found there were some weird parts in inheritance. I don't have enough knowledge in this layer yet to make a decision, so let's discuss the right interface. (or it could be follow-up)

Reply:
Looking into it further, it seems that `device=None` is only for v1 block manager. In that case, I would have two versions of this function: `get_num_free_blocks_v1` without `device` parameter, and `get_num_free_blocks` with `device` parameter. Then it becomes more obvious which code is using the old version of `get_num_free_blocks`; we could migrate that code to use v2 and then delete the unused v1 code.

Reply:
Sure. I like the suggestion. But let me ping @cadedaniel before final changes

Reply:
Thanks for making this PR @rkooo567 . Some details that may help:
* The `BlockAllocator` shouldn't have any device-aware details; if there are any device parts of the contract it should go into `DeviceAwareBlockAllocator`. The latter was made to inherit from the former to reduce code but seems it's adding to confusion, we should define the interfaces separately. We should be able to remove `BlockAllocator` device references after this. Seems I missed `allocate_immutable`'s device in a PR somehow -- it shouldn't take a device.
* FYI all block manager v1 code lives independently of the block manager v2 stuff, with the goal being complete deletion of it after v2 is feature complete

Reply:
Python doesn't support method overloading, and when I searched the current way is the recommended solution to support it https://stackoverflow.com/questions/10202938/how-do-i-use-method-overloading-in-python. 

My suggestion is to keep it in a way I did and remove it when we delete V1 (given it will be removed in a month or so). Wdyt?

Reply:
alternatively we can use different method name `allocate_mutable_v1` (as suggested) or `allocate_mutable_device`. But given we will remove this code soon anyway, it seems unnecessary?

Reply:
So the V1 is actually not involved here.

You can remove this issue by simply defining the DeviceAwareBlockManager interface without inheritance:
```python3
class DeviceAwareBlockAllocator(ABC):

    @abstractmethod
    def allocate_mutable(self, prev_block: Optional[Block],
                         device: Device) -> Block:
        pass
    @abstractmethod
    def allocate_immutable(self, prev_block: Optional[Block],
                           token_ids: List[int], device: Device) -> Block:
        pass

    @abstractmethod
    def free(self, block: Block) -> None:
        pass

    @abstractmethod
    def fork(self, last_block: Block) -> List[Block]:
        pass

    @abstractmethod
    def get_num_free_blocks(self, device: Device) -> int:
        pass

    @abstractmethod
    def clear_copy_on_writes(self) -> Dict[int, List[int]]:
        pass

    @abstractmethod
    def mark_blocks_as_computed(self) -> None:
        pass

    @abstractmethod
    def get_common_computed_block_ids(
            self, seq_block_ids: List[List[int]]) -> List[int]:
        pass
```

Reply:
fixed